File: scripts/verify_button_interactions.py
import json
import os
from playwright.sync_api import sync_playwright

BASE_URL = os.environ.get("E2E_BASE_URL", "http://localhost:3000")
results = []


def inspect_control(page, label, locator):
    locator.wait_for(state="visible")
    locator.scroll_into_view_if_needed()
    locator.focus()
    focus = locator.evaluate(
        """element => {
          const style = getComputedStyle(element);
          const rect = element.getBoundingClientRect();
          return {
            radius: style.borderRadius,
            outline: style.outline,
            outlineOffset: style.outlineOffset,
            focusVisible: element.matches(':focus-visible'),
            rect: {x: rect.x, y: rect.y, width: rect.width, height: rect.height},
          };
        }"""
    )
    rect = focus["rect"]
    page.mouse.move(rect["x"] + rect["width"] / 2, rect["y"] + rect["height"] / 2)
    hover = locator.evaluate(
        """element => {
          const style = getComputedStyle(element);
          return {hovered: element.matches(':hover'), transform: style.transform, backgroundColor: style.backgroundColor};
        }"""
    )
    if focus["radius"] != "15px":
        raise AssertionError(f"{label} radius was {focus['radius']}, expected 15px.")
    if not focus["focusVisible"] or focus["outline"] == "none":
        raise AssertionError(f"{label} did not present a visible keyboard focus style.")
    if not hover["hovered"]:
        raise AssertionError(f"{label} did not activate :hover.")
    results.append({"label": label, "focus": focus, "hover": hover})


def inspect_viewport(browser, name, viewport, mobile):
    context = browser.new_context(viewport=viewport, is_mobile=mobile, has_touch=mobile)
    page = context.new_page()
    try:
        page.goto(f"{BASE_URL}/", wait_until="networkidle")
        inspect_control(page, f"{name}: hero Check availability", page.get_by_role("link", name="Check availability", exact=True))
        inspect_control(page, f"{name}: booking Start over", page.get_by_role("button", name="Start over"))
        if not mobile:
            inspect_control(page, f"{name}: header Book Direct", page.get_by_role("link", name="Book Direct", exact=True).first)
        else:
            menu = page.get_by_role("button", name="Open navigation")
            if menu.count():
                inspect_control(page, f"{name}: navigation menu", menu)
        start_over = page.get_by_role("button", name="Start over")
        start_over.focus()
        page.keyboard.press("Enter")
        start_over.wait_for(state="visible")
        results.append({"label": f"{name}: Start over keyboard press", "passed": True})
    finally:
        context.close()


with sync_playwright() as playwright:
    browser = playwright.chromium.launch(headless=True)
    try:
        inspect_viewport(browser, "desktop", {"width": 1280, "height": 720}, False)
        inspect_viewport(browser, "mobile", {"width": 375, "height": 812}, True)
    finally:
        browser.close()

print(json.dumps({"passed": True, "results": results}, indent=2))

File: scripts/verify-button-interactions.mjs
import { chromium } from "playwright";

const baseUrl = process.env.E2E_BASE_URL ?? "http://localhost:3000";
const browser = await chromium.launch({ headless: true });
const results = [];

async function inspectControl(page, label, locator) {
  await locator.waitFor({ state: "visible" });
  await locator.focus();
  const focus = await locator.evaluate(element => {
    const style = getComputedStyle(element);
    const rect = element.getBoundingClientRect();
    return {
      radius: style.borderRadius,
      outline: style.outline,
      outlineOffset: style.outlineOffset,
      focusVisible: element.matches(":focus-visible"),
      rect: { x: rect.x, y: rect.y, width: rect.width, height: rect.height },
    };
  });

  const rect = focus.rect;
  await page.mouse.move(rect.x + rect.width / 2, rect.y + rect.height / 2);
  const hover = await locator.evaluate(element => {
    const style = getComputedStyle(element);
    return { hovered: element.matches(":hover"), transform: style.transform, backgroundColor: style.backgroundColor };
  });

  if (focus.radius !== "15px") throw new Error(`${label} radius was ${focus.radius}, expected 15px.`);
  if (!focus.focusVisible || focus.outline === "none") throw new Error(`${label} did not present a visible keyboard focus style.`);
  if (!hover.hovered) throw new Error(`${label} did not activate :hover.`);
  results.push({ label, focus, hover });
}

async function inspectViewport(name, viewport, isMobile) {
  const context = await browser.newContext({ viewport, isMobile, hasTouch: isMobile });
  const page = await context.newPage();
  await page.goto(`${baseUrl}/`, { waitUntil: "networkidle" });

  await inspectControl(page, `${name}: hero Check availability`, page.getByRole("link", { name: "Check availability", exact: true }));
  await inspectControl(page, `${name}: booking Start over`, page.getByRole("button", { name: /Start over/i }));

  if (!isMobile) {
    await inspectControl(page, `${name}: header Book Direct`, page.locator('header a[href="/booking"]').first());
  } else {
    const menuButton = page.getByRole("button", { name: /menu|navigation/i }).first();
    if (await menuButton.count()) await inspectControl(page, `${name}: navigation menu`, menuButton);
  }

  const startOver = page.getByRole("button", { name: /Start over/i });
  await startOver.focus();
  await page.keyboard.press("Enter");
  await startOver.waitFor({ state: "visible" });
  results.push({ label: `${name}: Start over keyboard press`, passed: true });
  await context.close();
}

try {
  await inspectViewport("desktop", { width: 1280, height: 720 }, false);
  await inspectViewport("mobile", { width: 375, height: 812 }, true);
  console.log(JSON.stringify({ passed: true, results }, null, 2));
} finally {
  await browser.close();
}
